### ISSUE TITLE
Bundled excon breaks things

### ISSUE THREAD

**Issue by mikesimons** [chef-cookbooks/docker#516]
Today I encountered a very strange issue where Excon was entering an infinite loop parsing headers when trying to use chef-provisioning with DigitalOcean (to provision a docker host).

The relevant part of the backtrace...
```
 "/home/mike/projects/cluster-scheduling/local-mode-cache/cache/cookbooks/docker/files/default/vendor/gems/excon-0.45.4/lib/excon/headers.rb:33:in `[]='",
 "/home/mike/projects/cluster-scheduling/local-mode-cache/cache/cookbooks/docker/files/default/vendor/gems/excon-0.45.4/lib/excon/headers.rb:33:in `[]='",
 "/home/mike/projects/cluster-scheduling/local-mode-cache/cache/cookbooks/docker/files/default/vendor/gems/excon-0.45.4/lib/excon/headers.rb:33:in `[]='",
 "/home/mike/projects/cluster-scheduling/local-mode-cache/cache/cookbooks/docker/files/default/vendor/gems/excon-0.45.4/lib/excon/headers.rb:33:in `[]='",
 "/home/mike/projects/cluster-scheduling/local-mode-cache/cache/cookbooks/docker/files/default/vendor/gems/excon-0.45.4/lib/excon/headers.rb:33:in `[]='",
 "/home/mike/projects/cluster-scheduling/local-mode-cache/cache/cookbooks/docker/files/default/vendor/gems/excon-0.45.4/lib/excon/headers.rb:33:in `[]='",
 "/home/mike/projects/cluster-scheduling/local-mode-cache/cache/cookbooks/docker/files/default/vendor/gems/excon-0.45.4/lib/excon/headers.rb:33:in `[]='",
 "/opt/chefdk/embedded/lib/ruby/gems/2.1.0/gems/excon-0.45.4/lib/excon/headers.rb:33:in `[]='",
 "/opt/chefdk/embedded/lib/ruby/gems/2.1.0/gems/excon-0.45.4/lib/excon/response.rb:172:in `parse_headers'",
 "/opt/chefdk/embedded/lib/ruby/gems/2.1.0/gems/excon-0.45.4/lib/excon/response.rb:75:in `parse'",
 "/opt/chefdk/embedded/lib/ruby/gems/2.1.0/gems/excon-0.45.4/lib/excon/middlewares/response_parser.rb:6:in `response_call'",
```

The vendored excon loop continues until a SystemStackError is raised.

I believe that this is because there are two Excon gems in play with the following (or something like it) timeline:

1. ChefDK Excon gem loaded
2. Vendored Excon gem loaded, Excon::Headers reopened
3. The reopening of the Excon::Headers causes the alias_method soup to point to the already aliased methods

The following snippet (based on excon/headers.rb) confirms:

```
> cat test.rb 
(1..2).each do
  class ReopenTest < Hash
    alias_method :raw_writer, :[]=
    def []=(key, value)
      raw_writer(key, value)
      puts "#{key}:#{value}"
    end
  end
  x = ReopenTest.new
  x[:a] = :b
end

> ruby test.rb
a:b
test.rb:5: stack level too deep (SystemStackError)
```

Is there a specific reason for vendoring these gems as opposed to using `chef_gem 'excon'` or such?

**Comment by someara** [comment]
Hi. docker-api is vendored because the docker_container and docker_image resources rely on it... 
excon is simply a dependency.

I'm going to close this for now due to non-actionability on my side... can you re-open this on the provisioning repo and link here so people finding this via google can follow it?

Thanks!

-s

**Comment by kri5** [comment]
I got the same issue using chef-provisionning. Btw there is no reason for you to vendor you gems dependencies, and you shoud use chef_gem resource, because that leads into conflicts with the 'same' gem being loaded from different location. That's why everybody uses chef_gem resource to install gem, then requires it, so that if multiple cookbooks need the same gem, they won't load it multiple times, as they are loaded in the same context.

**Comment by someara** [comment]
paging @jkeiser

**Comment by kri5** [comment]
Any updates?

**Comment by someara** [comment]
@jkeiser, you mentioned a library loading trick we could use to get around this?

**Comment by kri5** [comment]
Still got the issue, do you have any ideas for a fix and/or a workaround?

**Comment by kri5** [comment]
I do not know you're setup, but i think it's maybe related to the fact i use chef-dk as @mikesimons does, and chef-dk ruby install packages excon too.

**Comment by mikesimons** [comment]
`chef_gem` or something like https://github.com/inviqa/chef-config-driven-helper/blob/master/recipes/mysql-ruby.rb#L4 is the only solution that is viable where a cookbook depends on gems in my experience. `chef_gem` gets evaluated immediately so if gem install depends on something from your cookbook run (like mysql libs) then you may need to resort to the `gem_install` specifying the ChefDK gem binary method which will cause the gem to be installed in-order. AFAIK neither docker nor excon have such dependencies so `chef_gem` should be safe.

excon has proved problematic because it is written in such a way the double evaluation changes behaviour but you will also encounter problems with half-and-half gems. This is evident in my backtrace where it starts off using the ChefDK excon gem but then switches to the vendored gem. Can you imagine the seven shades of fun that might occur if these gem versions do not match! :grinning:

@kri5 I abandonded chef-provisioning for the project from which I encountered this so I'm afraid I have no workaround sans forking and implementing as I suggest above.

**Comment by someara** [comment]
Made some modifications to autoload... can somebody let me know if this an issue in v2.3.11?

**Comment by someara** [comment]
This should be fixed up in 2.3.15 via the new _autoload
